File: python-soa/order-service/app/infrastructure/web/dto.py
"""
Order DTOs (Data Transfer Objects)
====================================
Pydantic models for the Order Service API.
"""

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class OrderItemRequest(BaseModel):
    """Request body for a single order item."""

    product_id: str = Field(..., description="Product UUID")
    product_name: str = Field(default="", description="Product name for display")
    quantity: int = Field(..., gt=0, description="Quantity to order")
    unit_price: float = Field(..., gt=0, description="Price per unit")


class CreateOrderRequest(BaseModel):
    """Request body for creating a new order."""

    user_id: str = Field(..., description="User UUID placing the order")
    items: list[OrderItemRequest] = Field(
        ..., min_length=1, description="Order items (at least one required)"
    )


class CancelOrderRequest(BaseModel):
    """Request body for cancelling an order."""

    reason: str = Field(default="User requested cancellation", description="Cancellation reason")


class OrderItemResponse(BaseModel):
    """Response body for a single order item."""

    id: UUID
    product_id: UUID
    product_name: str
    quantity: int
    unit_price: float
    subtotal: float


class OrderResponse(BaseModel):
    """Response body representing an order."""

    id: UUID
    user_id: UUID
    status: str
    total_amount: float
    items: list[OrderItemResponse]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: nodejs-soa/order-service/src/infrastructure/web/dto.ts
// =============================================================================
// Data Transfer Objects (DTOs) - Order Service
// =============================================================================

/**
 * Request body for creating an order.
 */
export interface CreateOrderDto {
  userId: string;
  items: CreateOrderItemDto[];
}

export interface CreateOrderItemDto {
  productId: string;
  productName: string;
  quantity: number;
  unitPrice: number;
}

/**
 * Response DTO for an order.
 */
export interface OrderResponseDto {
  id: string;
  userId: string;
  status: string;
  totalAmount: number;
  items: OrderItemResponseDto[];
  createdAt: string;
  updatedAt: string;
}

export interface OrderItemResponseDto {
  id: string;
  productId: string;
  productName: string;
  quantity: number;
  unitPrice: number;
  subtotal: number;
}

/**
 * Validates the create order request body.
 */
export function validateCreateOrderDto(body: any): string | null {
  if (!body.userId || typeof body.userId !== 'string') {
    return 'userId is required and must be a string';
  }
  if (!body.items || !Array.isArray(body.items) || body.items.length === 0) {
    return 'items is required and must be a non-empty array';
  }

  for (let i = 0; i < body.items.length; i++) {
    const item = body.items[i];
    if (!item.productId || typeof item.productId !== 'string') {
      return `items[${i}].productId is required and must be a string`;
    }
    if (!item.productName || typeof item.productName !== 'string') {
      return `items[${i}].productName is required and must be a string`;
    }
    if (!item.quantity || typeof item.quantity !== 'number' || item.quantity <= 0) {
      return `items[${i}].quantity is required and must be a positive number`;
    }
    if (!item.unitPrice || typeof item.unitPrice !== 'number' || item.unitPrice <= 0) {
      return `items[${i}].unitPrice is required and must be a positive number`;
    }
  }

  return null;
}
